src/hotspot/cpu/x86/macroAssembler_x86.cpp

```diff
@@ -4912,6 +4912,10 @@ void MacroAssembler::population_count(Register dst, Register src,
     }
     bind(done);
   }
+#ifdef ASSERT
```

Comment:
I understand that it greatly increases the reproducibility of the bug, but does such an adhoc code snippet adds much from maintenance perspective? Reusing registers is common in assembly code, so if all places where a register value is destroyed are annotated in a similar way, it would add a lot of noise on source code level.

Reply:
I'm looking at `MacroAssembler::lookup_secondary_supers_table_var()` and the only thing which can be improved there is an extra comment around line 5116 stressing that `population_count` destroys `slot`.

Reply:
> I understand that it greatly increases the reproducibility of the bug, but does such an adhoc code snippet adds much from maintenance perspective? Reusing registers is common in assembly code, so if all places where a register value is destroyed are annotated in a similar way, it would add a lot of noise on source code level.

The point isn't so much that it's destroyed, but that it's destroyed on a rarely-used code path, in this case on a code path that is only ever executed on an obsolete machine. We've had the same problem on AArch64, where there are the old and new versions of compare-and-swap, and fixed the testing in the same way.

Reply:
Ok, thanks for the pointer. 

I'm fine with keeping the patch as it is, but I would definitely prefer a cleaner way to zap registers when it comes to scaling its usage across the code base.

FTR there's a special  place `globalDefinitions.hpp` for constants used for debugging. And I'd be confused to find out that the constant is spelled in camel case after trying to grep for it :-)

Reply:
It might be useful to have a testing flag that simulates old hardware on modern hardware.  So for example, it would turn off UsePopCountInstruction.  We have something similar on x86 with UseKNLSetting.

Reply:
Doesn't `-XX:UseSSE=2` achieve exactly that on x86-64?

Reply:
> Doesn't `-XX:UseSSE=2` achieve exactly that on x86-64?

Mmm, but x86 so many overlapping architecture changes over the years it's impossible to test every combination of them.

What would be good, though, would be to run tests on whatever the minimum architecture is that we support. I don't think we have any equivalent of  what GCC calls -march=x86-64, "a generic CPU with 64-bit extensions, but it looks like -XX:UseSSE=2 means Opteron aka k8, which "supersets MMX, SSE, SSE2, 3DNow!, and enhanced 3DNow!" and is our minimum CPU.

Reply:
It's definitely not ideal. And the main limitation of VM flag-only solution is that it can't catch bugs when unsupported instructions are issued (ISA capability checks being too lax). So, can't fully substitute testing on a real hardware, but getting pretty close IMO.